### ISSUE TITLE
[suse leap] Replication failure

### ISSUE THREAD

**Issue by Marenz** [rockstor/rockstor-core#2215]
When replicating a share from one instance to another, 
if it once failed to reach the other instance:
```
Receiver(icebear:10002) is unreachable. Aborting.. Exception: Receiver(icebear:10002) is unreachable. Aborting.
```
it will from then on fail with every new attempt saying 
```
Failed to create snapshot: video_1_replication_2. Aborting.. Exception: 500 Server Error: INTERNAL SERVER ERROR for url: http://127.0.0.1:8000/api/shares/4/snapshots/video_1_replication_2
```
From what I gather this is because that snapshot with that name already exists? After I deleted it manually, it worked again.
I suspect the previous failed attempt created a new snapshot to be transferred and didn't properly clean it up when it failed?

**Comment by phillxnet** [comment]
@Marenz Thanks for the issue creation.

> From what I gather this is because that snapshot with that name already exists?

Yes that's exactly it. This is a long standing issue unfortunately. And not a 'Built on openSUSE' specific one. We had the exact same behaviour in our CentOS Stable channel where replication was fixed having been broken previously (by API changes) in our CentOS based testing channel.

> After I deleted it manually, it worked again.

And that is the work around currently.

> I suspect the previous failed attempt created a new snapshot to be transferred and didn't properly clean it up when it failed?

Yes. We are simply a little starved of developer time to tend to this bug. And given it's in the replication system it can be a time consuming function to fully test. But given it's fairly well understood to be as you have described it, it may not be that ardous to test this specific fix. But to then run through the 5 cycles it takes for our replication to 'settle' can still take some time.

I thought we already had an issue open for this actually but I can't find it right now. We do have the more upstream fragility that can upset replication:

https://github.com/rockstor/rockstor-core/issues/1106

But what you have detailed here is hopefully something we improve upon independently.

Thanks again for the nicely focused issue. For cases where the receiver is just not available, we might fairly easily address this failure by just doing a quick test of access to receive before we proceed. And then log accordingly. This way we can at least guard against this simple stumbling block on our part in a 'machine not available' scenarios. And hopefully as the btrfs send/receive subsystem gets it's next revision soon we will get more failure modes that we can catch appropriately. But the most obvious fragility on our part is that we trip ourselves up before even check on the receive's availability. And so addressing this first would allow for a far more robust setup that we currently have given target machine availability is likely not a given in may circumstances, and blocking all future replication events is definitely shooting ourselves in the foot. I'm just conscious of trying to be careful in what is already a fragile arrangement re many btrfs send/receive failure possibilities. I.e. we also don't account for partial send failure. But we could, for the time being catch the instance you describe I think.

Cheers.

**Comment by Hooverdan96** [comment]
Once there is an opportunity to revisit this challenge (as there are a couple of more recent forum posts around cleaning up a failure and to proceed, maybe these two can be an inspiration on how to approach a retry/failure clean up, etc.:

https://github.com/nachoparker/btrfs-sync and

https://github.com/digint/btrbk

**Comment by phillxnet** [comment]
@Hooverdan96 Linking to a comment of mine from last year in:

"Implement Replication failure detection and recovery for a btrfs corner case" #1106

quoting the author of btrbk: https://github.com/rockstor/rockstor-core/issues/1106#issuecomment-308419962 on the fail we also are affected by re underlying limitations of btrfs send/receive, which btrbk uses; but from what I can tell, btrfs-sync does not. Only just had a quick look at that btrfs-sync however.

But note a later comment there where we discussed this same failure, and alternatives.

We likely want to close all related issues and combine in a new issue the various references:
https://github.com/rockstor/rockstor-core/issues/1106#issuecomment-1862499660

**Comment by Hooverdan96** [comment]
yes, I originally wanted to create a new issue but then thought about searching for other replication issues (since I couldn't remember anything anymore). So, agreed.

**Comment by phillxnet** [comment]
@Hooverdan96 @FroggyFlox I'm pretty sure, for this particular instance/issue, we have a currently remarked out safeguard that simply needs to be re-implemented / brought up-to-date with a change in our API that lead to it's being remarked out.

That is, again from memory, I'm pretty sure we use-to skip a replication event if the receiving system was not found. Ergo removing the show-stopper that we now run into in this instance where we fail a little later on when the sending snapshot has already been created.

Again, we just need more eyes on this code, and there are now more comments there since my last visit regarding such things.
I.e. likely some existing TODO's that may not take that much attention to resolve.

So in this issue's specific case:

> From what I gather this is because that snapshot with that name already exists?

we already have a older safeguard that simply needs to be re-enabled. So lets leave this issue open to address just that instance of our current failure. I.e. it does not directly represent an in-process btrfs-send event failure. But our own wrapper failing to first ensure the receivings station is in place.

Above needs verifying I'm afraid, as currently working in a very different ares of the code,  but just wanted to clarify that we do have some currently in-active safeguards that were in the original design that are currently yet to be updated to deep API change from a few years ago now.

**Comment by Hooverdan96** [comment]
> @Hooverdan96 @FroggyFlox I'm pretty sure, for this particular instance/issue, we have a currently remarked out safeguard that simply needs to be re-implemented / brought up-to-date with a change in our API that lead to it's being remarked out.

I think this is the Issue and its fix you were referring to:
https://github.com/rockstor/rockstor-core/issues/1853

and its associated PR: https://github.com/rockstor/rockstor-core/pull/1885

may be this piece that was commented out?
https://github.com/rockstor/rockstor-core/blob/1ddcf4b6f6ad6a451fdaef492fe974417d4dbfe3/src/rockstor/smart_manager/replication/receiver.py#L251-L258
